linalg/sundials.hpp

```diff
@@ -108,6 +119,15 @@ class CVODESolver : public ODESolver, public SundialsSolver
    static int LinSysSolve(SUNLinearSolver LS, SUNMatrix A, N_Vector x,
                           N_Vector b, realtype tol);
 
+   static int root(realtype t, N_Vector y, realtype *gout, void *user_data);
+
+   typedef std::function<int(realtype t, Vector y, Vector gout, CVODESolver *)>
```

Comment:
I'm ok with it, but I want people with stronger feelings about such things to comment on the inclusion of `std::function`.

Reply:
@tzanio Any comments about the use of std::function here?

Reply:
Can you explain why you prefer that to a function pointer?

From consistency perspective, I will prefer to keep the same interface as in the rest of the library, i.e. a function pointer.

However if there are significant benefits, I'm okay with including it here...

Reply:
The main benefits to using std::function is having the ability to capture using lambdas as opposed to just using standard function pointers.

For example: 
https://github.com/mfem/mfem/blob/343f9e02a607338f14d228d8f8b17ad2fdb2c404/miniapps/adjoint/cvsRoberts_ASAi_dns.cpp#L177-L189

A user would be able to specify `reltol`, `abstol_v` and you would not have to create a new function and create some "global" variables to pass the tolerances around.

Reply:
Ok, lets keep this.

For backward compatibility though, can you add versions that use function pointers? The users can decide how to pass additional parameters to those.

Reply:
The interface is already compatible with a raw function pointer from what I can tell.

Reply:
It should be backwards compatible with raw function pointers also.